File: VB/Scaffolding.ManyToMany/ViewModels/Course/CourseViewModel.vb
Imports System
Imports System.Linq
Imports System.Linq.Expressions
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports Scaffolding.ManyToMany.Common.Utils
Imports Scaffolding.ManyToMany.SchoolContextDataModel
Imports Scaffolding.ManyToMany.Common.DataModel
Imports Scaffolding.ManyToMany.Model
Imports Scaffolding.ManyToMany.Common.ViewModel

Namespace Scaffolding.ManyToMany.ViewModels
    ''' <summary>
    ''' Represents the single Course object view model.
    ''' </summary>
    Partial Public Class CourseViewModel
        Inherits SingleObjectViewModel(Of Course, Integer, ISchoolContextUnitOfWork)

        ''' <summary>
        ''' Creates a new instance of CourseViewModel as a POCO view model.
        ''' </summary>
        ''' <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        Public Shared Function Create(Optional ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of ISchoolContextUnitOfWork) = Nothing) As CourseViewModel
            Return ViewModelSource.Create(Function() New CourseViewModel(unitOfWorkFactory))
        End Function

        ''' <summary>
        ''' Initializes a new instance of the CourseViewModel class.
        ''' This constructor is declared protected to avoid undesired instantiation of the CourseViewModel type without the POCO proxy factory.
        ''' </summary>
        ''' <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        Protected Sub New(Optional ByVal unitOfWorkFactory As IUnitOfWorkFactory(Of ISchoolContextUnitOfWork) = Nothing)
            MyBase.New(If(unitOfWorkFactory, UnitOfWorkSource.GetUnitOfWorkFactory()), Function(x) x.Courses, Function(x) x.Title)
        End Sub
    End Class
End Namespace

File: CS/Scaffolding.ManyToMany/ViewModels/Course/CourseViewModel.cs
using System;
using System.Linq;
using System.Linq.Expressions;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;
using DevExpress.Mvvm.DataModel;
using DevExpress.Mvvm.ViewModel;
using Scaffolding.ManyToMany.SchoolContextDataModel;
using Scaffolding.ManyToMany.Common;
using Scaffolding.ManyToMany.Model;

namespace Scaffolding.ManyToMany.ViewModels {

    /// <summary>
    /// Represents the single Course object view model.
    /// </summary>
    public partial class CourseViewModel : SingleObjectViewModel<Course, int, ISchoolContextUnitOfWork> {

        /// <summary>
        /// Creates a new instance of CourseViewModel as a POCO view model.
        /// </summary>
        /// <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        public static CourseViewModel Create(IUnitOfWorkFactory<ISchoolContextUnitOfWork> unitOfWorkFactory = null) {
            return ViewModelSource.Create(() => new CourseViewModel(unitOfWorkFactory));
        }

        /// <summary>
        /// Initializes a new instance of the CourseViewModel class.
        /// This constructor is declared protected to avoid undesired instantiation of the CourseViewModel type without the POCO proxy factory.
        /// </summary>
        /// <param name="unitOfWorkFactory">A factory used to create a unit of work instance.</param>
        protected CourseViewModel(IUnitOfWorkFactory<ISchoolContextUnitOfWork> unitOfWorkFactory = null)
            : base(unitOfWorkFactory ?? UnitOfWorkSource.GetUnitOfWorkFactory(), x => x.Courses, x => x.Title) {
        }


        protected override void RefreshLookUpCollections(bool raisePropertyChanged) {
            base.RefreshLookUpCollections(raisePropertyChanged);
            StudentsDetailEntities = CreateAddRemoveDetailEntitiesViewModel(x => x.Students, x => x.Students);
        }

        public virtual AddRemoveDetailEntitiesViewModel<Course, Int32, Student, Int32, ISchoolContextUnitOfWork> StudentsDetailEntities { get; protected set; }
    }
}
